array.h

```diff
@@ -2487,6 +2495,11 @@ class array {
   operator const_reference() const {
     return *base_;
   }
+
+  template <typename NewShape, typename T2, typename Shape2, typename Alloc2>
```

Comment:
Nit: I'd call it `ShapeOut`, `TOut`, ...

Reply:
Isn't NewShape the ShapeOut?

This is annoying because we need new template parameters that don't shadow T, Shape, Alloc, they are otherwise the same (except Shape, which could be called OldShape).